### DISCUSSION TITLE
Custom server: Can't conditionally serve different Next directories

### DISCUSSION THREAD

**Discussion by just-learning-1** [vercel/next.js#33552] [Help]
### Run `next info` (available from version 12.0.8 and up)

    Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 20.5.0: Sat May  8 05:10:33 PDT 2021; root:xnu-7195.121.3~9/RELEASE_X86_64
    Binaries:
      Node: 16.13.2
      npm: 8.1.2
      Yarn: 1.22.17
      pnpm: N/A
    Relevant packages:
      next: 12.0.8
      react: 17.0.2
      react-dom: 17.0.2

### What version of Next.js are you using?

12.0.8

### What version of Node.js are you using?

16.13.2

### What browser are you using?

Safari

### What operating system are you using?

macOS

### How are you deploying your application?

node server.js

### Describe the Bug

I prepared a minimal repro and issue description here: [Next.js no conditional server issue (minimal repro)](https://github.com/just-learning-1/minimal-repros/tree/master/nextjs-no-conditional-server-issue)

### Goal

I want to conditionally serve different Next directories to different sessions within a custom Express server.

### Issue Summary

After Next app server serves one directory (`next({ dir: 'a' })`) in a server runtime, it seems impossible to later serve a different directory (`next({ dir: 'b' })`), even to an entirely different session or machine.

### Expected behavior

(See code & repro) I expected to get served `dir: 'a'` when storing cookie value `'a'`, and to get served `dir: 'b'` when storing cookie value `'b'`.

Instead, after getting served either `'a'` or `'b'` once, all future attempts to conditionally serve will serve that first directory.

## Repro

1.  Clone the [minimal repro](https://github.com/just-learning-1/minimal-repros/tree/master/nextjs-no-conditional-server-issue)
2.  Run `yarn serve`
3.  Visit `localhost:4000` in a private browser window
4.  Read: `Visit /a or /b to continue`
5.  Visit `localhost:4000/a` (for this example)
6.  Read: `Your choice is a. Now, return to /`
7.  Visit `localhost:4000`
8.  Read: `Welcome to a`
9.  Now, visit `localhost:4000` in a different private browser window (or different machine)) and read instruction
10. Visit `localhost:4000/b` and read message
11. Finally, visit `localhost:4000`

**Expected:** Read: `Welcome to b`

**Actual:** Read: `Welcome to a`

*Note:* Visiting `/b` first will indeed persist `Welcome to b` instead.

## Code

Here's the code in `server.js`, minimally changed from the official [Custom Express Server example](https://github.com/vercel/next.js/tree/canary/examples/custom-server-express):

```js
// server.js
const cookieParser = require('cookie-parser')
const express = require('express')
const next = require('next')

const port = parseInt(process.env.PORT, 10) || 4000
const dev = process.env.NODE_ENV !== 'production'

// Want different next apps & handlers to conditionally source different directories
const appA = next({ dev, dir: 'a' })
const handleA = appA.getRequestHandler()
const appB = next({ dev, dir: 'b' })
const handleB = appB.getRequestHandler()

const server = express()

server.use(cookieParser())

// Visit /a to receive cookie with choice=a
server.get('/a', (req, res) => {
  res.cookie('choice', 'a')
  res.status(200).send('Your choice is a. Now, return to /')
})

// Visit /b to receive cookie with choice=b
server.get('/b', (req, res) => {
  res.cookie('choice', 'b')
  res.status(200).send('Your choice is b. Now, return to /')
})

server.all('*', (req, res, goNext) => {
  const cookieChoice = req.cookies['choice'] || ''
  const chooseMessage = 'Visit /a or /b to continue'

  if (!cookieChoice) {
    // When no cookie, send message
    res.send(chooseMessage)
  } else {
    let app
    let handle
    switch (cookieChoice) {
      case 'a':
        // Set app & handler to serve directory 'a/'
        app = appA
        handle = handleA
        break
      case 'b':
        // Set app & handler to serve directory 'b/'
        app = appB
        handle = handleB
        break
      default:
        res.send(chooseMessage)
    }

    // Want to conditionally serve 'a/' or 'b/' here
    app.prepare()
      .then(() => {
        handle(req, res) 
      })

    goNext()
  }
})

server.listen(port, (err) => {
  if (err) throw err
  console.log(`> Ready on http://localhost:${port}`)
})
```

**Comment by zhangwei900808** [comment]
same problem! only us see this???

**Comment by zhangwei900808** [comment]
A cookie appear B chrome brower but B cookie must NULL!!!

**Comment by wwwdepot** [comment]
Are you trying to do AB testing? You can accomplish that using middleware & it [supports custom server](https://nextjs.org/docs/middleware#custom-server).

Middleware AB testing example: https://github.com/vercel/examples/tree/main/edge-functions/ab-testing-simple
